libraries/AP_GPS/AP_GPS_SBP2.cpp

```diff
@@ -0,0 +1,494 @@
+void
+AP_GPS_SBP2::_sbp_process_message() {
+
+    //Here, we copy messages into local structs.
+    switch (parser_state.msg_type) {
+        case SBP_HEARTBEAT_MSGTYPE:
+            last_heartbeat = *((struct sbp_heartbeat_t*)parser_state.msg_buff);
+            last_heartbeat_received_ms = AP_HAL::millis();
+            waiting_for_first_heartbeat = false;
+            break;
+
+        case SBP_GPS_TIME_MSGTYPE:
+            last_gps_time = *((struct sbp_gps_time_t*)parser_state.msg_buff);
+            break;
+
+        case SBP_VEL_NED_MSGTYPE:
+            last_vel_ned = *((struct sbp_vel_ned_t*)parser_state.msg_buff);
+            break;
+
+        case SBP_POS_LLH_MSGTYPE:
+            last_pos_llh = *((struct sbp_pos_llh_t*)parser_state.msg_buff);
+            break;
+
+        case SBP_DOPS_MSGTYPE:
+            last_dops = *((struct sbp_dops_t*)parser_state.msg_buff);
+            break;
+
+        default:
+            break;
+    }
+
+    // send all messages we receive to log, even if it's an unsupported message,
+    // so we can do annditional post-processing from Dataflash logs.
+    // The log mask will be used to adjust or suppress logging
+    logging_log_raw_sbp(parser_state.msg_type, parser_state.sender_id, parser_state.msg_len, parser_state.msg_buff);
+}
+
+bool
+AP_GPS_SBP2::_attempt_state_update()
+{
+
+    uint32_t now = AP_HAL::millis();
+
+    if (waiting_for_first_heartbeat)
+        return false;
+
+    if (now - last_heartbeat_received_ms > SBP_TIMEOUT_HEARTBEAT) {
+
+        state.status = AP_GPS::NO_GPS;
+        Debug("No Heartbeats from Piksi! NO GPS, Dying.");
+        return false;
+
+    } else if (last_heartbeat.protocol_major < 2) {
+
+        state.status = AP_GPS::NO_GPS;
+        Debug("Received a heartbeat from a SBPv1 device. Current driver does not support this protocol. NO GPS, Dying. ");
+        return false;
+
+    } else if (last_heartbeat.nap_error   == 1 ||
+               last_heartbeat.io_error    == 1 ||
+               last_heartbeat.sys_error   == 1) {
+
+        state.status = AP_GPS::NO_GPS;
+
+        Debug("Piksi reported an error. Printing Health Report. NO GPS, Dying.");
+        Debug("   ext_antenna: %d", last_heartbeat.ext_antenna);
+        Debug("   res2: %d", last_heartbeat.res2);
+        Debug("   protocol_major: %d", last_heartbeat.protocol_major);
+        Debug("   protocol_minor: %d", last_heartbeat.protocol_minor);
+        Debug("   res: %d", last_heartbeat.res);
+        Debug("   nap_error: %d", last_heartbeat.nap_error);
+        Debug("   io_error: %d", last_heartbeat.io_error);
+        Debug("   sys_error: %d", last_heartbeat.sys_error);
+
+        return false;
+
+    } else if (last_pos_llh.tow == last_vel_ned.tow
+            && abs((int32_t) (last_gps_time.tow - last_vel_ned.tow)) < 10000
+            && abs((int32_t) (last_dops.tow - last_vel_ned.tow)) < 60000
+            && last_vel_ned.tow > last_full_update_tow) {
```

Comment:
oh dear, it's even worse than that - if the week wraps, this driver will never do another update. this is leftover from the previous driver, let me fix both. great catch, thank you!